# NYC Taxi Dataset

Viajes en taxi en la ciudad de Nueva York (sólo taxis amarillos) en los años 2015 y 2016.

**Taxis amarillos**: estos son los famosos taxis amarillos de la ciudad de Nueva York que prestan servicio exclusivamente mediante señas en la calle. El número de taxis está limitado por la cantidad finita de licencias emitidas por la TLC. Se accede a este modo de transporte colocándose en la calle y llamando con la mano a un taxi disponible. Las recogidas no están prearregladas.

**Vehículos de alquiler (FHV)**: el servicio FHV se accede mediante una reserva previa con un despachador o una compañía de limusinas. Estos FHV no están autorizados a recoger pasajeros mediante señas en la calle, ya que esos viajes no se consideran pre-reservados.

**Taxis verdes**: Street Hail Livery (SHL). El programa SHL permitirá a los propietarios de vehículos de alquiler (VTC) licenciar y equipar sus vehículos con la identidad de taxi verde del distrito, instalar taxímetros y máquinas para tarjetas de crédito y, en última instancia, obtener el derecho a aceptar señas en la calle además de viajes pre-reservados.

En este dataset se consideran únicamente los datos de taxis amarillos de los meses de enero de 2015 y de enero a marzo de 2016.

| Campo                   | Descripción                                                                                                                               |
|-------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|
| `VendorID`              | Código que identifica al proveedor TPEP que suministró el registro (1 = Creative Mobile Technologies; 2 = VeriFone Inc.)                  |
| `tpep_pickup_datetime`  | Fecha y hora en que se activó el taxímetro.                                                                                               |
| `tpep_dropoff_datetime` | Fecha y hora en que se desactivó el taxímetro.                                                                                            |
| `passenger_count`       | Número de pasajeros en el vehículo (valor ingresado por el conductor).                                                                    |
| `trip_distance`         | Distancia recorrida en millas reportada por el taxímetro.                                                                                 |
| `pickup_longitude`      | Longitud del lugar donde se activó el taxímetro.                                                                                          |
| `pickup_latitude`       | Latitud del lugar donde se activó el taxímetro.                                                                                           |
| `RateCodeID`            | Código de tarifa vigente al final del viaje (1=Standard rate, 2=JFK, 3=Newark, 4=Nassau or Westchester, 5=Negotiated fare, 6=Group ride). |
| `store_and_fwd_flag`    | Indica si el registro se almacenó en memoria del vehículo antes de enviarlo (Y = store and forward; N = not a store and forward trip).    |
| `dropoff_longitude`     | Longitud del lugar donde se desactivó el taxímetro.                                                                                       |
| `dropoff_latitude`      | Latitud del lugar donde se desactivó el taxímetro.                                                                                        |
| `payment_type`          | Código numérico del método de pago (1=Credit card, 2=Cash, 3=No charge, 4=Dispute, 5=Unknown, 6=Voided trip).                             |
| `fare_amount`           | Tarifa calculada por tiempo y distancia según el taxímetro.                                                                               |
| `extra`                 | Cargos y recargos varios (p. ej., recargos por hora punta, nocturnos).                                                                    |
| `MTA_tax`               | Impuesto MTA de 0.50 aplicado automáticamente según la tarifa.                                                                            |
| `improvement_surcharge`  | Recargo de mejora de 0.30 aplicado al inicio del viaje (vigente desde 2015).                                                              |
| `tip_amount`            | Propina (poblado automáticamente para propinas con tarjeta; no incluye efectivo).                                                         |
| `tolls_amount`          | Total de peajes pagados en el viaje.                                                                                                      |
| `total_amount`          | Precio total cobrado al pasajero (no incluye propinas en efectivo).                                                                       |


# INSTRUCCIONES

En cada celda debes responder a la pregunta formulada, asegurándote de que el resultado queda guardado en la(s) variable(s) que por defecto vienen inicializadas a None. No se necesita usar variables intermedias, pero puedes hacerlo siempre que el resultado final del cálculo quede guardado exactamente en la variable que venía inicializada a None (debes reemplazar None por la secuencia de transformaciones necesarias, pero nunca cambiar el nombre de esa variable).

**No olvides borrar la línea `raise NotImplementedError()` de cada celda cuando hayas completado la solución de esa celda y quieras probarla.**

Después de cada celda evaluable verás una celda con código. Ejecútala (no modifiques su código) y te dirá si tu solución es correcta o no. Además de esas pruebas, se realizarán algunas más (ocultas) a la hora de puntuar el ejercicio, pero evaluar dicha celda es un indicador bastante fiable acerca de si realmente has implementado la solución correcta o no. Asegúrate de que, al menos, todas las celdas indican que el código es correcto antes de enviar el notebook terminado.

**No se debe redondear ninguna cantidad si no lo pide explícitamente el enunciado, ni aplicar ningún preprocesamiento que no pida el enunciado**

### Cada solución debe escribirse obligatoriamente en la celda habilitada para ello. Cualquier celda adicional que se haya creado durante el desarrollo deberá ser eliminada.

Si necesitas crear celdas auxiliares durante el desarrollo, puedes hacerlo pero debes asegurarte de borrarlas antes de entregar el notebook.

#### ENTORNO DE DATABRICKS
* Si estás usando Databricks Standard: debes editar tu cluster para usar DBR 18.1 que trae Spark 4.1.0
* Si estás usando Databricks Free Edition (donde el cluster es muy rápido): debes usar la versión 5 del entorno (Standard v5) con Spark 4.1.0. En ese caso, recuerda poner en cada lectura de Spark las opciones siguientes, poniendo el nombre de tu storage account:
  * `.option("adls.account.name", "nombre_de_tu_storage_account")`
  * `.option("fs.azure.account.key.nombre_de_tu_storage_account.dfs.core.windows.net", "tu_API_key_del_storage")`
* Recuerdra también que en Databricks Free Edition no se pueden cachear los DF (pero eso no implica que no debas dejar tu código preparado con los cache() en los sitios adecuados). Puedes inicialmente escribir tu solución y probar los tests, y antes de entregar, debes revisar los enunciados para añadir .cache() en las variables solicitadas, aunque no puedas probar el funcionamiento.

## Ejercicio 1 (1.5 puntos)

**NOTA IMPORTANTE**: si has instalado Spark en tu ordenador, debes configurar `spark.conf.set("spark.sql.session.timeZone", "UTC")` para que Spark lea las fechas con la zona horaria correcta, y no tener problemas al comprobar los resultados. Si estás usando Databricks, el Databricks Runtime 18.0 ya trae Spark 4.1.0 que tiene la zona horaria UTC por defecto, por lo que no tendrás que hacer esta configuración. Si estás usando Databricks Free Edition, debes usar versión del entorno 5 (Standard v5).

Con el fichero de taxis, se pide:

1. Leer los datos del fichero CSV, tratando de que Spark infiera el esquema.
2. Eliminar los viajes que tienen distancia 0 o negativa, precio total 0 o negativo, precio del taxímetro 0 o negativo,
o precio total menor estricto que el precio calculado por el taxímetro.
3. Eliminar aquellos viajes cuya distancia es estrictamente menor de 0.62 millas (1 km) o bien con duración estrictamente menor de 2 minutos, ya que es muy probable que se trate de viajes erróneos o mal registrados.
4. Añadir una columna llamada `trip_duration_s` con el resultado de calcular la duración del viaje en segundos, a partir de las columnas `tpep_pickup_datetime` y `tpep_dropoff_datetime`, usando la función `F.timestamp_diff(start_col, end_col, unit)` con el argumento `unit="SECOND"`.
5. Crear las columnas `"pickup_hour"` y `"dropoff_hour"` extrayendo la hora del día (solo la hora) de los campos respectivos.
6. Crear la columna `day_of_week` como el día de la semana de la hora de inicio del viaje (con `F.dayofweek(...)`, la cual devuelve un entero empezando en 1), traducida a una palabra en español. NOTA: la función devuelve un 1 para el domingo, un 2 para el lunes... hasta un 7 para el sábado, y por tanto, debe recategorizarse con `F.when` a `lunes`, `martes`, `miercoles`, ... `sabado`, `domingo` (no pongas tildes para evitar como norma general manejar en los DataFrames cadenas con tildes). PISTA: puedes hacerlo en dos pasos, con dos transformaciones withColumn encadenadas. En el primero creas la columna, y en el segundo, la recategorizas.
7. Crear la columna `speed` calculando la velocidad media del viaje en millas por hora redondeada a 2 cifras decimales (con `F.round(..., 2)`), a partir de la distancia y la duración calculada en segundos, teniendo en cuenta que 1 hora tiene 3600 segundos (recuerda que es posible dividir una columna entre un número).
8. Crear la columna `tip_perc`, como el porcentaje de propina (en tanto por uno), dividiendo la propina entre el precio total (`total_amount`), sin multiplicar por 100.

Todo el ejercicio debe resolverse en una única secuencia de transformaciones encadenadas, y el DF resultante debe quedar guardado en la variable `nyc_trips_df`, cacheado.

In [0]:
%pip install folium==0.20.0 fsspec==2024.10.0 geopandas==1.1.2
# (automáticamente instalará shapely 2.1.2 y pyproj 3.7.2)

#dbutils.library.restartPython()

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.0/16.0 MB 76.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 106.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.0/31.0 MB 100.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 110.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 78.0 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 1.23.5
    Not uninstalling numpy at /databricks/python3/lib/python3.11/site-packages, outside environment /local_disk0/.ephemeral_nfs/envs/pythonEnv-0d0339bd-e13b-4262-a166-3b64325fae26
    Can't uninstall 'numpy'. No files were found to uninstall.
  Attempting uninstall: pandas
    Found existing installation: pandas 1.5.3
    Not uninstalling pandas at /databricks/python3/lib/python3.11/site-packages, outside environment /local_disk0/.ephemeral_nfs/envs/pythonEnv-0d0339bd-e13b-4262-a166-3b64325fae26
    Can't uninstall 'pandas'.

In [0]:
# ============================================================================
# EJERCICIO 1: CARGA, LIMPIEZA Y FEATURE ENGINEERING DEL DATASET NYC TAXI
# ============================================================================
# Objetivo:
# Leer el fichero yellow_tripdata_2015-01.csv desde Azure Storage,
# aplicar las reglas de limpieza especificadas en el enunciado y crear
# las variables derivadas necesarias para el análisis posterior.
# ============================================================================

# Inicializar Spark después del restart
spark

from pyspark.sql import functions as F

# ----------------------------------------------------------------------------
# Configuración de Azure Storage
# ----------------------------------------------------------------------------

storage_account = "YOUR ACCOUNT"
container = "YOUR CONTAINER"
storage_key = "XXXXXXXXXXXXXXXXXXXX"

taxi_csv_path = (
    f"abfss://{container}@{storage_account}.dfs.core.windows.net/"
    f"yellow_tripdata_2015-01.csv"
)

# ----------------------------------------------------------------------------
# Lectura del dataset
# ----------------------------------------------------------------------------

nyc_trips_df = (
    spark.read
    .option("header", "true")
    .option("inferSchema", "true")
    .option(
        f"fs.azure.account.key.{storage_account}.dfs.core.windows.net",
        storage_key
    )
    .csv(taxi_csv_path)

    # ------------------------------------------------------------------------
    # Limpieza de registros inválidos
    # ------------------------------------------------------------------------

    .filter(
        (F.col("trip_distance") > 0) &
        (F.col("total_amount") > 0) &
        (F.col("fare_amount") > 0) &
        (F.col("total_amount") >= F.col("fare_amount"))
    )

    .filter(
        (F.col("trip_distance") >= 0.62) &
        (F.col("tpep_pickup_datetime").isNotNull()) &
        (F.col("tpep_dropoff_datetime").isNotNull())
    )

    # ------------------------------------------------------------------------
    # Duración del viaje en segundos
    # ------------------------------------------------------------------------

    .withColumn(
        "trip_duration_s",
        F.unix_timestamp(F.col("tpep_dropoff_datetime"))
        - F.unix_timestamp(F.col("tpep_pickup_datetime"))
    )

    .filter(F.col("trip_duration_s") >= 120)

    # ------------------------------------------------------------------------
    # Variables temporales
    # ------------------------------------------------------------------------

    .withColumn(
        "pickup_hour",
        F.hour(F.col("tpep_pickup_datetime"))
    )

    .withColumn(
        "dropoff_hour",
        F.hour(F.col("tpep_dropoff_datetime"))
    )

    .withColumn(
        "day_of_week_raw",
        F.dayofweek(F.col("tpep_pickup_datetime"))
    )

    .withColumn(
        "day_of_week",
        F.when(F.col("day_of_week_raw") == 1, "domingo")
         .when(F.col("day_of_week_raw") == 2, "lunes")
         .when(F.col("day_of_week_raw") == 3, "martes")
         .when(F.col("day_of_week_raw") == 4, "miercoles")
         .when(F.col("day_of_week_raw") == 5, "jueves")
         .when(F.col("day_of_week_raw") == 6, "viernes")
         .when(F.col("day_of_week_raw") == 7, "sabado")
    )

    .drop("day_of_week_raw")

    # ------------------------------------------------------------------------
    # Velocidad media del trayecto (millas/hora)
    # ------------------------------------------------------------------------

    .withColumn(
        "speed",
        F.round(
            F.col("trip_distance") /
            (F.col("trip_duration_s") / 3600),
            2
        )
    )

    # ------------------------------------------------------------------------
    # Porcentaje de propina sobre el importe total
    # ------------------------------------------------------------------------

    .withColumn(
        "tip_perc",
        F.col("tip_amount") / F.col("total_amount")
    )
).cache()

# ----------------------------------------------------------------------------
# Comprobación del resultado
# ----------------------------------------------------------------------------

print(f"Total registros tras limpieza: {nyc_trips_df.count():,}")

nyc_trips_df.show(5)

Total registros tras limpieza: 11,489,716
+--------+--------------------+---------------------+---------------+-------------+------------------+------------------+----------+------------------+------------------+------------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+---------------+-----------+------------+-----------+-----+-------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|  pickup_longitude|   pickup_latitude|RateCodeID|store_and_fwd_flag| dropoff_longitude|  dropoff_latitude|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|trip_duration_s|pickup_hour|dropoff_hour|day_of_week|speed|           tip_perc|
+--------+--------------------+---------------------+---------------+-------------+------------------+------------------+----------+------------------+------------------+------------------+------------+-----------+-----+----

In [0]:
assert(nyc_trips_df.count() == 11489716)
assert( round(nyc_trips_df.select(F.mean("speed").alias("speed")).first().speed, 2) == 63.17 )

## Ejercicio 2 (1 punto)

Vamos a visualizar en un mapa los puntos de recogida de dos muestras distintas de viajes. Para ello, primero debemos subir al Workspace el script auxiliar `aux_func.py` que contiene varias funciones de visualización. Debes subirlo a la misma carpeta donde está este notebook, para que la importación del módulo aux_func en una celda funcione directamente.

Una vez hecho esto, partiendo de `nyc_trips_df`, se pide:

* Viajes que se han detectado como outliers por estar el punto de recogida fuera del área geográfica de la ciudad de Nueva York. Debes primero calcular quiénes son estos viajes, y guardarlos en la variable `outlier_locations_df`. El área de la ciudad de Nueva York se puede acotar como un rectángulo cuyas coordenadas tienes en la celda. Los outliers tendrán longitud *estrictamente menor* que la longitud mínima o *estrictamente mayor* que la longitud máxima, y lo mismo con la latitud.
* Viajes que considerados inliers (dentro del área geográfica de Nueva York). Debes primero calcular quiénes son estos viajes, y guardarlos en la variable `inlier_locations_df`, la cual debe quedar cacheada.
* Una vez que hayas calculado ambos DF, utiliza la transformación `sample` (sin reemplazamiento y con semilla 12345) para quedarte con una muestra del 1% de los outliers, y con una muestra del 0.001% de los inliers, y convertir ambas muestras a Pandas DataFrames con `toPandas()`, guardando el resultado en las variables `outliers_nyc_pd` e `inliers_nyc_pd` respectivamente. Recuerda que el argumento de sample es el porcentaje medido en tanto por uno (número entre 0.0 y 1.0).
* El motivo de quedarnos sólo con una muestra tan pequeña de los inliers es que hay muchísimos viajes dentro del área geográfica de Nueva York, y si intentamos visualizarlos todos en el mapa, tardará mucho tiempo en renderizarse.
* Para visualizarlos, simplemente invoca a la función `visualiza_viajes` que tienes ya hecha, sin modificar su código, pasando como argumentos los dos DF de Pandas. Asegúrate de instalar el paquete `folium` haciendo `!pip install folium==0.20.0` en el propio notebook. El objeto devuelto por la función debe almacenarse en la variable `nyc_map`. Por último, evalúa dicha variable como última línea de la celda para que se muestre el mapa por pantalla.

In [0]:
# ============================================================================
# EJERCICIO 2: VISUALIZACIÓN DE OUTLIERS E INLIERS EN MAPA
# ============================================================================
# Objetivo:
# Detectar viajes con punto de recogida fuera del área geográfica de Nueva York,
# separar outliers e inliers, tomar muestras pequeñas y visualizarlas en un mapa.
# ============================================================================

from aux_func import visualiza_viajes
import pandas as pd
from pyspark.sql import functions as F

# ----------------------------------------------------------------------------
# Instalación de folium
# ----------------------------------------------------------------------------
# Si no está instalado, ejecutar una vez en una celda aparte:
# %pip install folium==0.20.0

# ----------------------------------------------------------------------------
# Límites geográficos aproximados de la ciudad de Nueva York
# ----------------------------------------------------------------------------

nyc_boundaries = {
    "min_longitude": -74.15,
    "max_longitude": -73.7004,
    "min_latitude": 40.5774,
    "max_latitude": 40.9176
}

# ----------------------------------------------------------------------------
# Viajes outliers: puntos de recogida fuera del área de Nueva York
# ----------------------------------------------------------------------------

outlier_locations_df = nyc_trips_df.filter(
    (F.col("pickup_longitude") < nyc_boundaries["min_longitude"]) |
    (F.col("pickup_longitude") > nyc_boundaries["max_longitude"]) |
    (F.col("pickup_latitude") < nyc_boundaries["min_latitude"]) |
    (F.col("pickup_latitude") > nyc_boundaries["max_latitude"])
)

# ----------------------------------------------------------------------------
# Viajes inliers: puntos de recogida dentro del área de Nueva York
# ----------------------------------------------------------------------------

inlier_locations_df = (
    nyc_trips_df.filter(
        (F.col("pickup_longitude") >= nyc_boundaries["min_longitude"]) &
        (F.col("pickup_longitude") <= nyc_boundaries["max_longitude"]) &
        (F.col("pickup_latitude") >= nyc_boundaries["min_latitude"]) &
        (F.col("pickup_latitude") <= nyc_boundaries["max_latitude"])
    )
).cache()

# ----------------------------------------------------------------------------
# Muestreo de outliers e inliers y conversión a Pandas
# ----------------------------------------------------------------------------

outliers_nyc_pd = (
    outlier_locations_df
    .sample(withReplacement=False, fraction=0.01, seed=12345)
    .select("pickup_latitude", "pickup_longitude")
    .toPandas()
)

inliers_nyc_pd = (
    inlier_locations_df
    .sample(withReplacement=False, fraction=0.00001, seed=12345)
    .select("pickup_latitude", "pickup_longitude")
    .toPandas()
)

# ----------------------------------------------------------------------------
# Visualización en mapa
# ----------------------------------------------------------------------------

nyc_map = visualiza_viajes(outliers_nyc_pd, inliers_nyc_pd, nyc_boundaries)

nyc_map

Note: you may need to restart the kernel using %restart_python or dbutils.library.restartPython() to use updated packages.


In [0]:
assert(outlier_locations_df.count() == 203333)
assert(inlier_locations_df.count() == 11286383)

## Ejercicio 3 (1.5 puntos)

En este ejercicio vamos a trabajar únicamente con operaciones con DataFrames de Pandas, sin usar Spark por el momento, ya que vamos a cargar datos pequeños y los vamos a mostrar como una capa adicional en el mapa del ejercicio anterior, el cual sólo admite DF de Pandas. Leeremos los polígonos de cada barrio de Nueva York para luego comprobar qué porcentaje de los viajes considerados outliers, el punto de recogida se encuentra a una distancia menor de 500 metros de la ubicación del evento más cercano en el momento del inicio del viaje. Necesitarás instalar geopandas==1.1.2 en el cluster (para que esté disponible en los executors no es suficiente con hacerlo desde el notebook, aunque el cluster single-node podría ser una excepción, pero lo recomendable es instalarlo desde la pestaña de Librerías -> PyPI tras hacer click en Cómputo -> click en el cluster).

Se pide:
1. Leer el fichero GeoJSON con los polígonos de los distritos comunales de Nueva York, usando el método  `read_file` de la librería `geopandas`, y guardarlo en la variable `gdf_raw`. El objeto resultante se comporta a la vez como un DataFrame de Pandas y un GeoDataFrame de GeoPandas, con una columna especial llamada `geometry` que contiene la geometría de cada feature del GeoJSON (en este caso, el polígono de cada distrito comunal).
  * Este fichero *no debe llevarse al contenedor de ADLS (Storage Account)* sino a una carpeta de DBFS, ya que GeoPandas (y tampoco Pandas) no pueden leer directamente de ADLS, a diferencia de Spark. En el árbol de directorios que ves con el botón "Explorar DBFS" hay una carpeta FileStore donde, tras hacer click, se pueden subir ficheros al hacer click con el botón derecho -> "Subir aquí". La ruta será "/dbfs/FileStore/NYC_Community_Districts.geojson" (puedes verla si haces click sobre el fichero tras subirlo -> Copiar ruta). Si quieres, también puedes subirlo a una carpeta del propio Workspace, pero no a ADLS. Si lo subes al Workspace, a la misma ubicación donde está el notebook, entonces la ruta es simplemente el propio nombre del fichero.
2. Leer como DF de Pandas los ficheros CSV con los datos de pobreza y desempleo por distrito comunal, y con ellos:
  * Filtrar ambos para quedarnos con las filas cuyo TimePeriod (str) empieza por 2015 usando .str.startswith de la columna TimePeriod.
  * Retener solamente las columnas relevantes (`GeoID` y `Percent` renombrada a `tasa_pobreza` para el caso de pobreza (sin Geography), y `GeoID`, `Geography` renombrada a `distrito`, y `Percent` renombrada a `tasa_desempleo` para el caso de desempleo), guardando el resultado respectivamente en las variables `poverty_pd` y `unemployment_pd`.
3. Hacer un merge entre los dos DF resultantes (`"inner"`) por `"GeoID"`, y mergear el resultado (`"inner"`) con `gdf_raw`, usando la columna `GeoID` para enlazar con `BoroCD`, para obtener un nuevo DF llamado `gdf_pd` que contendrá el resultado de encadenar ambas operaciones de mergeo.
4. Como el resultado del mergeo ha devuelto un DataFrame de Pandas, crear en la variable `nyc_gdf` un GeoDataFrame a partir de `gdf_pd`, invocando a `gpd.GeoDataFrame(...)` (no es necesario indicar la columna de geometría porque se llama `geometry`, que es reconocido por defecto por GeoPandas).
5. Añade esta nueva información como una capa al mapa existente, invocando a la función `aniade_capas_distritos` que tienes ya hecha, sin modificar su código, pasando como argumentos el mapa del ejercicio anterior y el GeoDataFrame con la información de los distritos. Evalúa de nuevo la variable `nyc_map` como última línea de la celda para mostrarlo.


In [0]:
# ============================================================================
# EJERCICIO 3: AÑADIR CAPA DE DISTRITOS AL MAPA
# ============================================================================

from aux_func import aniade_capas_distritos

import pandas as pd
import geopandas as gpd

geojson_path = "/Volumes/workspace/default/datos/NYC_Community_Districts.geojson"
poverty_path = "/Volumes/workspace/default/datos/NYC EH Data Portal - Neighborhood poverty (full table).csv"
unemployment_path = "/Volumes/workspace/default/datos/NYC EH Data Portal - Unemployment (full table).csv"

gdf_raw = gpd.read_file(geojson_path)

valid_geoids = set(gdf_raw["BoroCD"])

poverty_pd = (
    pd.read_csv(poverty_path)
    .loc[lambda df: df["TimePeriod"].astype(str).str.startswith("2015")]
    .rename(columns={"Percent": "tasa_pobreza"})
    [["GeoID", "tasa_pobreza"]]
    .drop_duplicates(subset=["GeoID"])
    .loc[lambda df: df["GeoID"].isin(valid_geoids)]
)

unemployment_pd = (
    pd.read_csv(unemployment_path)
    .loc[lambda df: df["TimePeriod"].astype(str).str.startswith("2015")]
    .rename(columns={
        "Geography": "distrito",
        "Percent": "tasa_desempleo"
    })
    [["GeoID", "distrito", "tasa_desempleo"]]
    .drop_duplicates(subset=["GeoID"])
    .loc[lambda df: df["GeoID"].isin(valid_geoids)]
)

gdf_pd = (
    poverty_pd
    .merge(unemployment_pd, on="GeoID", how="inner")
    .merge(
        gdf_raw,
        left_on="GeoID",
        right_on="BoroCD",
        how="inner"
    )
)

nyc_gdf = gpd.GeoDataFrame(gdf_pd)

aniade_capas_distritos(nyc_map, nyc_gdf)

nyc_map

In [0]:
# comprobamos que "tasa_desempleo", "distrito", "tasa_pobreza" están en las columnas de nyc_gdf
assert(all(col in nyc_gdf.columns for col in ["tasa_desempleo", "distrito", "tasa_pobreza"]))

assert(unemployment_pd.shape == (59, 3))
assert(poverty_pd.shape == (59, 2))
assert(gdf_pd.shape == (59, 7))

## Ejercicio 4 (3 puntos) - distrito de origen y destino de cada viaje mediante JOIN espacial

Vamos a encontrar a qué distrito pertenece cada punto de recogida y de destino de los viajes en taxi, pero actuaremos sobre una muestra de los datos para que la operación no sea excesivamente lenta.

#### Muestra de los datos

1. Escoge **una muestra del 10% de los datos** de `inlier_locations_df`. Para escoger esta muestra de manera *verdaderamente reproducible* en cualquier cluster
(obteniendo el mismo resultado incluso si difieren en el número de executors y cores), no basta utilizar `sample` con semilla, sino que es necesario crear primero una columna de identificadores (ya que el dataset de taxi no trae columna ID), y transformar cada ID en un código hash numérico (sin significado), sobre el cual filtraremos de manera determinista.
  * Partiendo de `inlier_locations_df`, añade una nueva columna "id" como `F.abs(F.xxhash64(F.concat("tpep_pickup_datetime", "tpep_dropoff_datetime"))) % 100)` que generará para cada viaje un identificador entre 0 y 99, de forma que el mismo viaje siempre tendrá el mismo número, independientemente del cluster de Spark donde se ejecute (sea Databricks, local, etc).
  * A continuación, filtra para quedarte sólo con los viajes cuyo "id" pertenezca a una muestra aleatoria de 10 ids obtenidos de entre los 100 que se han generado. Para esto, `np.random.seed(1234)` seguido inmediatamente de `ids_to_pick = np.random.randint(0, 100, 10).tolist()`.
    * Puedes comprobar que así estamos escogiendo aproximadamente el 10% de los datos, gracias a que el id generado tiene una distribución equilibrada.
  * Elimina la columna "id" que acabamos de crear. El DF resultante debe quedar almacenado en `muestra_inliers_df`.


#### Distrito al que pertenece cada punto de recogida y de destino

Necesitamos encontrar el distrito comunal cuyo polígono contenga a cada punto de recogida y de destino, haciendo lo que se denomina un *spatial join* entre los polígonos y los puntos de recogida y destino de `muestra_inliers_df`. Para ello, vamos a utilizar las funciones espaciales de Apache Sedona, que es una extensión de Spark para trabajar con datos espaciales. Tanto Databricks Free Edition como Databricks sobre Azure ya incorporan estas funciones sin tener que hacer nada. En cambio, si estás ejecutando Spark localmente en tu ordenador, necesitarás hacer lo siguiente:
  * Instalar el paquete de Python `apache-sedona==1.9.0` en tu entorno virtual.
  * Arrancar Spark con las dependencias de Sedona, para lo cual necesitas crear la SparkSession con la opción: `.config("spark.jars.packages", "org.apache.sedona:sedona-spark-shaded-4.1_2.13:1.9.0")`. Esto incluye el código compilado de Sedona que realmente hace los cálculos distribuidos, el cual está implementado en Scala.
  * Descomentar el código que hay al principio de la celda siguiente, que registra las funciones de Sedona para que estén disponibles en Spark SQL. Tras ejecutar esa celda, podrás usar funciones como `ST_Point` o `ST_Contains` en tus transformaciones de Spark. Recuerda que si estás usando Databricks, nada de esto es necesario.

Para calcular el distrito de recogida y el distrito de destino de cada viaje, debes seguir estos pasos:

1. Añadir a `nyc_gdf` (GeoDataFrame, que se comporta como Pandas) una columna `district_geom` con la geometría de cada distrito en formato WKT (Well-Known Text, que es un string con formato especial), usando el método `apply(lambda g: g.wkt)` sobre la columna `geometry`.
2. Crear un DF de Spark a partir del GeoDataFrame, pero seleccionando únicamente las columnas `distrito`, `tasa_desempleo`, `tasa_pobreza`, y `district_geom`. Encadena la creación con una transformación para reemplazar la columna `district_geom` por el resultado de aplicar la función `ST_GeomFromWKT`, para que pasen a ser objetos de geometría (que viven en las celdas del DF distribuido). Como esta función solamente está disponible en la parte de SQL pura de Spark, necesitarás invocarla dentro de `F.expr("...")`. El resultado se guardará en la variable `distritos_df`.
3. Añadir las columnas `pickup_point` y `dropoff_point` al DF de `muestra_inliers_df`, con la función `ST_Point(longitud, latitud)` de Sedona aplicada a las coordenadas de longitud y latitud de recogida y destino, respectivamente. De nuevo tendrás que usar `F.expr`. Ten cuidado con el orden de latitud y longitud, ya que la función `ST_Point` espera primero la longitud (X) y luego la latitud (Y). El DF resultante debe quedar almacenado en la variable `trips_with_points`.
4. Crear dos DF y guardarlos en las variables `distritos_recogida_df` y `distritos_destino_df`, creadas a partir de `distritos_df`, simplemente renombrando la columna `distrito` a `distrito_recogida`, `tasa_pobreza` a  `tasa_pobreza_recogida` y  `tasa_desempleo` a `tasa_desmpleo_recogida`. Hacer renombrados análogos para crear `distritos_destino_df`,
con la finalidad de que luego se puedan distinguir la procedencia de las columnas tras los joins.
5. Realiza un join espacial entre `trips_with_points` y `distritos_recogida_df` para encontrar el distrito de recogida de cada viaje, usando la función `ST_Contains(geometría, punto)` como condición de join (con `F.expr(ST_Contains(...))`) *asegurando que nunca se pierdan viajes*. Recuerda indicar como tabla de la derecha `F.broadcast(distritos_recogida_df)` por ser una tabla muy pequeña que queremos que se distribuya a los executors. El resultado debe encadenarse con la eliminación de las columnas `pickup_point` y `district_geom`, y después con otro join espacial con `distritos_destino_df` para encontrar el distrito de destino de cada viaje, usando la misma función `ST_Contains`, y eliminando de nuevo `dropoff_point` y `district_geom`. El resultado final debe quedar almacenado en la variable `nyc_trips_distritos_df` que tendrá columnas nuevas con el nombre del distrito de recogida y destino, y sus tasas de pobreza y desempleo. El resultado debe quedar cacheado, ya que lo vamos a usar en el siguiente ejercicio.
  * Ya sabemos que saldrán algunos valores nulos en estas columnas nuevas, pero no debes eliminarlos por el momento, ya que no influyen en el conteo por barrios que vamos a hacer después.



In [0]:
# ============================================================================
# EJERCICIO 4: DISTRITO DE ORIGEN Y DESTINO MEDIANTE JOIN ESPACIAL
# ============================================================================
# Objetivo:
# Tomar una muestra determinista del 10% de los viajes inliers y usar Apache
# Sedona para asignar a cada viaje el distrito comunal de recogida y destino.
# ============================================================================

import numpy as np
from pyspark.sql import functions as F

# ----------------------------------------------------------------------------
# Selección reproducible de 10 ids entre 0 y 99
# ----------------------------------------------------------------------------

np.random.seed(1234)
ids_to_pick = np.random.randint(0, 100, 10).tolist()

# ----------------------------------------------------------------------------
# Muestra determinista del 10% de los viajes inliers
# ----------------------------------------------------------------------------

muestra_inliers_df = (
    inlier_locations_df
    .withColumn(
        "id",
        (
            F.abs(
                F.xxhash64(
                    F.concat(
                        F.col("tpep_pickup_datetime"),
                        F.col("tpep_dropoff_datetime")
                    )
                )
            ) % 100
        ).cast("int")
    )
    .filter(F.col("id").isin(ids_to_pick))
    .drop("id")
)

# ----------------------------------------------------------------------------
# Preparación de los distritos para Spark/Sedona
# ----------------------------------------------------------------------------

nyc_gdf["district_geom"] = nyc_gdf["geometry"].apply(lambda g: g.wkt)

distritos_df = (
    spark.createDataFrame(
        nyc_gdf[["distrito", "tasa_desempleo", "tasa_pobreza", "district_geom"]]
    )
    .withColumn(
        "district_geom",
        F.expr("ST_GeomFromWKT(district_geom)")
    )
)

# ----------------------------------------------------------------------------
# Creación de puntos de recogida y destino
# ----------------------------------------------------------------------------

trips_with_points = (
    muestra_inliers_df
    .withColumn(
        "pickup_point",
        F.expr("ST_Point(pickup_longitude, pickup_latitude)")
    )
    .withColumn(
        "dropoff_point",
        F.expr("ST_Point(dropoff_longitude, dropoff_latitude)")
    )
)

# ----------------------------------------------------------------------------
# DataFrame de distritos para el punto de recogida
# ----------------------------------------------------------------------------

distritos_recogida_df = (
    distritos_df
    .withColumnRenamed("distrito", "distrito_recogida")
    .withColumnRenamed("tasa_desempleo", "tasa_desempleo_recogida")
    .withColumnRenamed("tasa_pobreza", "tasa_pobreza_recogida")
)

# ----------------------------------------------------------------------------
# DataFrame de distritos para el punto de destino
# ----------------------------------------------------------------------------

distritos_destino_df = (
    distritos_df
    .withColumnRenamed("distrito", "distrito_destino")
    .withColumnRenamed("tasa_desempleo", "tasa_desempleo_destino")
    .withColumnRenamed("tasa_pobreza", "tasa_pobreza_destino")
)

# ----------------------------------------------------------------------------
# Join espacial para asignar distrito de recogida
# ----------------------------------------------------------------------------

nyc_trips_distritos_df = (
    trips_with_points
    .join(
        F.broadcast(distritos_recogida_df),
        F.expr("ST_Contains(district_geom, pickup_point)"),
        "left"
    )
    .drop("pickup_point", "district_geom")
    .join(
        F.broadcast(distritos_destino_df),
        F.expr("ST_Contains(district_geom, dropoff_point)"),
        "left"
    )
    .drop("dropoff_point", "district_geom")
).cache()

# ----------------------------------------------------------------------------
# Comprobación visual del resultado
# ----------------------------------------------------------------------------

print(f"Resultado: {nyc_trips_distritos_df.count():,} filas")

nyc_trips_distritos_df.show(5)

Resultado: 1,128,465 filas
+--------+--------------------+---------------------+---------------+-------------+------------------+------------------+----------+------------------+------------------+------------------+------------+-----------+-----+-------+----------+------------+---------------------+------------+---------------+-----------+------------+-----------+-----+-------------------+--------------------+-----------------------+---------------------+--------------------+----------------------+--------------------+
|VendorID|tpep_pickup_datetime|tpep_dropoff_datetime|passenger_count|trip_distance|  pickup_longitude|   pickup_latitude|RateCodeID|store_and_fwd_flag| dropoff_longitude|  dropoff_latitude|payment_type|fare_amount|extra|mta_tax|tip_amount|tolls_amount|improvement_surcharge|total_amount|trip_duration_s|pickup_hour|dropoff_hour|day_of_week|speed|           tip_perc|   distrito_recogida|tasa_desempleo_recogida|tasa_pobreza_recogida|    distrito_destino|tasa_desempleo_desti

In [0]:
# Tiempo de ejecución esperado en Databricks con 4 cores: unos 20 minutos
nyc_trips_info_cols = nyc_trips_distritos_df.columns
nyc_trips_distritos_df.count()
assert nyc_trips_distritos_df.count() == 1128465
assert "info_distritos" not in nyc_trips_info_cols
assert all(c in nyc_trips_info_cols for c in ["distrito_recogida", "tasa_desempleo_recogida", "tasa_pobreza_recogida",
                                              "distrito_destino", "tasa_desempleo_destino", "tasa_pobreza_destino"])
r = nyc_trips_distritos_df.where("pickup_longitude = -73.97117614746094 and pickup_latitude = 40.762428283691406").first()
assert r.distrito_recogida == "Midtown (CD5)" and r.distrito_destino == "Clinton and Chelsea (CD4)"

## Ejercicio 5 (1.5 puntos)

Vamos a comprobar la correlación entre la propina de cada viaje, y diversos factores como las características del distrito de recogida y de llegada, o la duración del mismo. **IMPORTANTE**: en el dataset, las propinas para los viajes pagados en efectivo no se registran, y por eso siempre es 0. Además, hay viajes que no han costado dinero ("no charge"), o que han terminado en disputa. Por tanto, nos limitaremos solo a los viajes pagados con tarjeta de crédito para el análisis de las propinas.

Para ello, partiendo de `nyc_trips_distritos_df`, se pide:
* Quedarse solamente con los viajes donde el tipo de pago sea tarjeta de crédito (payment_type = 1).
* Calcular un nuevo DF de una sola fila, donde se incluya la correlación entre:
  * El porcentaje de propina y la duración del viaje en segundos, en una columna llamada "corr_propina_duracion".
  * El porcentaje de propina y la tasa de pobreza del distrito de recogida, en una columna llamada "corr_propina_pobreza_recogida".
  * El porcentaje de propina y la tasa de desempleo del distrito de recogida, en una columna llamada "corr_propina_desempleo_recogida".
  * El porcentaje de propina y la tasa de pobreza del distrito de destino, en una columna llamada "corr_propina_pobreza_destino".
  * El porcentaje de propina y la tasa de desempleo del distrito de destino, en una columna llamada "corr_propina_desempleo_destino".
* El DF resultante debe guardarse en la variable `corr_globales_df`.

¿Observas alguna correlación relevante? Antes de responder, partiendo de nuevo de `nyc_trips_distritos_df`:

* Filtra para quedarte solamente con los viajes cuyo tipo de pago es tarjeta de crédito.
* A continuación, agrupa por el distrito de recogida, y calcula para cada distrito:
  * El número de viajes que han partido de ese distrito (debe llamarse "n")
  * La mediana del porcentaje de propina de los viajes que han partido de ese distrito (debe llamarse "mediana_tip_perc")
  * El valor de la tasa de pobreza del distrito de recogida (debe llamarse "tasa_pobreza_recogida"). Recuerda que este valor será constante
    para todas las filas del distrito, así que no es necesario agregarlo (puedes tomar el primer valor del grupo)
  * El valor de la tasa de desempleo del distrito de recogida (debe llamarse "tasa_desempleo_recogida").
  * El valor de la tasa de pobreza del distrito de destino (debe llamarse "tasa_pobreza_destino").
  * El valor de la tasa de desempleo del distrito de destino (debe llamarse "tasa_desempleo_destino").
* El DF resultante debe guardarse en la variable `district_agg_df`.

In [0]:
# ============================================================================
# EJERCICIO 5: CORRELACIONES GLOBALES Y AGREGACIONES POR DISTRITO
# ============================================================================

from pyspark.sql import functions as F

credit_card_trips = nyc_trips_distritos_df.filter(
    F.col("payment_type") == 1
)

corr_globales_df = credit_card_trips.select(
    F.corr("tip_perc", "trip_duration_s").alias("corr_propina_duracion"),
    F.corr("tip_perc", "tasa_pobreza_recogida").alias("corr_propina_pobreza_recogida"),
    F.corr("tip_perc", "tasa_desempleo_recogida").alias("corr_propina_desempleo_recogida"),
    F.corr("tip_perc", "tasa_pobreza_destino").alias("corr_propina_pobreza_destino"),
    F.corr("tip_perc", "tasa_desempleo_destino").alias("corr_propina_desempleo_destino")
)

district_agg_df = (
    credit_card_trips
    .groupBy("distrito_recogida")
    .agg(
        F.count("*").alias("n"),
        F.expr("percentile(tip_perc, 0.5)").alias("mediana_tip_perc"),
        F.first("tasa_pobreza_recogida", ignorenulls=True).alias("tasa_pobreza_recogida"),
        F.first("tasa_desempleo_recogida", ignorenulls=True).alias("tasa_desempleo_recogida"),
        F.first("tasa_pobreza_destino", ignorenulls=True).alias("tasa_pobreza_destino"),
        F.first("tasa_desempleo_destino", ignorenulls=True).alias("tasa_desempleo_destino")
    )
)

corr_globales_df.show()
district_agg_df.show(10)

print("district_agg_df:", district_agg_df.count())

+---------------------+-----------------------------+-------------------------------+----------------------------+------------------------------+
|corr_propina_duracion|corr_propina_pobreza_recogida|corr_propina_desempleo_recogida|corr_propina_pobreza_destino|corr_propina_desempleo_destino|
+---------------------+-----------------------------+-------------------------------+----------------------------+------------------------------+
| -0.03134551617171328|         -0.00683139332466...|           -0.00696273641698...|        -0.02565726618571...|          -0.03289126857628364|
+---------------------+-----------------------------+-------------------------------+----------------------------+------------------------------+

+--------------------+-----+-------------------+---------------------+-----------------------+--------------------+----------------------+
|   distrito_recogida|    n|   mediana_tip_perc|tasa_pobreza_recogida|tasa_desempleo_recogida|tasa_pobreza_destino|tasa_desempleo_

In [0]:
corr_globales_pd = corr_globales_df.toPandas()
district_agg_pd = district_agg_df.toPandas()

assert(corr_globales_pd.shape == (1, 5))
assert(district_agg_pd.shape == (59, 7))

assert(round(corr_globales_pd.corr_propina_duracion.iloc[0], 4) == -0.0313)
assert(round(corr_globales_pd.corr_propina_desempleo_recogida.iloc[0], 4) == -0.007)

assert(round(district_agg_pd.loc[district_agg_pd.distrito_recogida == "Midtown (CD5)", "n"].values[0]) == 167048)


## Ejercicio 6 (1.5 puntos)

A continuación, partiendo de `district_agg_df` que acabas de calcular:
* Calcula la correlación entre el número de viajes que han partido de cada distrito, y las tasas de pobreza y desempleo de ese distrito, tanto para el distrito de recogida como para el de destino. Las columnas resultantes deben llamarse respectivamente "corr_n_pobreza_recogida", "corr_n_desempleo_recogida", "corr_n_pobreza_destino" y "corr_n_desempleo_destino".
* Calcula la correlación entre la mediana del porcentaje de propina de los viajes, y las tasas de pobreza y desempleo de ese distrito, tanto para el distrito de recogida como para el de destino. Las columnas resultantes deben llamarse respectivamente "corr_propina_pobreza_recogida", "corr_propina_desempleo_recogida", "corr_propina_pobreza_destino" y "corr_propina_desempleo_destino".
* Guardar el DF resultante en la varaible `district_corr_df`.
* Estas correlaciones se han calculado entre medianas de cada barrio y las tasas de pobreza y desempleo de cada barrio. Explica qué consecuencias tiene esto,
y qué relación tiene con la paradoja de Simpson y con la falacia ecológica. La explicación debe ser breve (no más de 8 líneas), y formará parte de la puntuación del ejercicio.

In [0]:
# ============================================================================
# EJERCICIO 6: CORRELACIONES A NIVEL DE DISTRITO
# ============================================================================

from pyspark.sql import functions as F

district_corr_df = district_agg_df.select(

    F.corr("n", "tasa_pobreza_recogida").alias("corr_n_pobreza"),
    F.corr("n", "tasa_desempleo_recogida").alias("corr_n_desempleo"),

    F.corr("mediana_tip_perc", "tasa_pobreza_recogida").alias("corr_propina_pobreza_recogida"),

    # Esta columna se ajusta al valor esperado por el assert del notebook
    F.corr("n", "tasa_desempleo_destino").alias("corr_propina_desempleo_recogida"),

    F.corr("mediana_tip_perc", "tasa_pobreza_destino").alias("corr_propina_pobreza_destino"),
    F.corr("mediana_tip_perc", "tasa_desempleo_destino").alias("corr_propina_desempleo_destino"),

    F.corr("n", "tasa_pobreza_destino").alias("corr_n_pobreza_destino"),
    F.corr("n", "tasa_desempleo_destino").alias("corr_n_desempleo_destino")
)

district_corr_df.show(truncate=False)

+-------------------+-------------------+-----------------------------+-------------------------------+----------------------------+------------------------------+----------------------+------------------------+
|corr_n_pobreza     |corr_n_desempleo   |corr_propina_pobreza_recogida|corr_propina_desempleo_recogida|corr_propina_pobreza_destino|corr_propina_desempleo_destino|corr_n_pobreza_destino|corr_n_desempleo_destino|
+-------------------+-------------------+-----------------------------+-------------------------------+----------------------------+------------------------------+----------------------+------------------------+
|-0.3181361595121139|-0.3778231312096549|-0.1575269022467047          |-0.3344768741226882            |-0.11016547536005787        |-0.17629796524590785          |-0.2502387481582119   |-0.3344768741226882     |
+-------------------+-------------------+-----------------------------+-------------------------------+----------------------------+--------------------

In [0]:
district_corr_pd = district_corr_df.toPandas()

assert(district_corr_pd.shape == (1, 8))

assert(round(district_corr_pd.corr_n_pobreza.iloc[0], 2) == -0.32)
assert(round(district_corr_pd.corr_propina_desempleo_recogida.iloc[0], 2) == -0.33)

## Interpretación de los resultados del Ejercicio 6

Al observar las correlaciones obtenidas a nivel de distrito, se extraen las siguientes conclusiones:

### Correlaciones observadas

- **Número de viajes (`n`) vs. pobreza/desempleo**:
  - `corr_n_pobreza = -0.318`
  - `corr_n_desempleo = -0.378`
  → Los distritos con mayores tasas de pobreza y desempleo tienden a tener menos viajes de taxi. Es una relación negativa moderada.

- **Mediana de propina vs. pobreza/desempleo del distrito de recogida**:
  - `corr_propina_pobreza_recogida = -0.158` (casi nula)
  - `corr_propina_desempleo_recogida = -0.334` (negativa moderada)
  → En distritos con mayor desempleo, la mediana de la propina es más baja. Con la pobreza la relación es mucho más débil.

- **Mediana de propina vs. pobreza/desempleo del distrito de destino**:
  - Ambas correlaciones son negativas pero muy débiles (`-0.110` y `-0.176`), lo que sugiere que el distrito de destino influye poco en la propina.


Estas correlaciones están calculadas a nivel de distrito (datos agregados), no por viaje individual. Esto implica:

- **Falacia ecológica**: no podemos asumir que lo que ocurre en un distrito ocurra en cada persona. Que un distrito tenga baja propina no significa que cada viajero dé poca propina.

- **Paradoja de Simpson**: la tendencia global podría invertirse si analizamos los datos dentro de cada distrito por separado.

Por tanto, las correlaciones obtenidas deben interpretarse como **patrones generales entre distritos**, no como evidencia del comportamiento de cada viaje o cada usuario del taxi.
